src/Access/LDAPAccessStorage.cpp

```diff
@@ -0,0 +1,309 @@
+    return memory_storage.findAll(type);
+}
+
+
+bool LDAPAccessStorage::existsImpl(const UUID & id) const
+{
+    std::scoped_lock lock(mutex);
+    return memory_storage.exists(id);
+}
+
+
+AccessEntityPtr LDAPAccessStorage::readImpl(const UUID & id) const
+{
+    std::scoped_lock lock(mutex);
+    return memory_storage.read(id);
+}
+
+
+String LDAPAccessStorage::readNameImpl(const UUID & id) const
+{
+    std::scoped_lock lock(mutex);
+    return memory_storage.readName(id);
+}
+
+
+bool LDAPAccessStorage::canInsertImpl(const AccessEntityPtr &) const
+{
+    return false;
+}
+
+
+UUID LDAPAccessStorage::insertImpl(const AccessEntityPtr & entity, bool)
+{
+    throwReadonlyCannotInsert(entity->getType(), entity->getName());
+}
+
+
+void LDAPAccessStorage::removeImpl(const UUID & id)
+{
+    std::scoped_lock lock(mutex);
+    auto entity = read(id);
+    throwReadonlyCannotRemove(entity->getType(), entity->getName());
+}
+
+
+void LDAPAccessStorage::updateImpl(const UUID & id, const UpdateFunc &)
+{
+    std::scoped_lock lock(mutex);
+    auto entity = read(id);
+    throwReadonlyCannotUpdate(entity->getType(), entity->getName());
+}
+
+
+ext::scope_guard LDAPAccessStorage::subscribeForChangesImpl(const UUID & id, const OnChangedHandler & handler) const
+{
+    std::scoped_lock lock(mutex);
+    return memory_storage.subscribeForChanges(id, handler);
+}
+
+
+ext::scope_guard LDAPAccessStorage::subscribeForChangesImpl(EntityType type, const OnChangedHandler & handler) const
+{
+    std::scoped_lock lock(mutex);
+    return memory_storage.subscribeForChanges(type, handler);
+}
+
+
+bool LDAPAccessStorage::hasSubscriptionImpl(const UUID & id) const
+{
+    std::scoped_lock lock(mutex);
+    return memory_storage.hasSubscription(id);
+}
+
+
+bool LDAPAccessStorage::hasSubscriptionImpl(EntityType type) const
+{
+    std::scoped_lock lock(mutex);
+    return memory_storage.hasSubscription(type);
+}
+
+UUID LDAPAccessStorage::loginImpl(const String & user_name, const String & password, const Poco::Net::IPAddress & address, const ExternalAuthenticators & external_authenticators) const
+{
+    std::scoped_lock lock(mutex);
+    auto id = memory_storage.find<User>(user_name);
+    if (id)
+    {
+        auto user = memory_storage.read<User>(*id);
+
+        if (!isPasswordCorrectImpl(*user, password, external_authenticators))
+            throwInvalidPassword();
+
+        if (!isAddressAllowedImpl(*user, address))
+            throwAddressNotAllowed(address);
+
+        return *id;
+    }
+    else
+    {
+        // User does not exist, so we create one, and will add it if authentication is successful.
+        auto user = std::make_shared<User>();
+        user->setName(user_name);
+        user->authentication = Authentication(Authentication::Type::LDAP_SERVER);
+        user->authentication.setServerName(ldap_server);
+
+        if (!isPasswordCorrectImpl(*user, password, external_authenticators))
+            throwInvalidPassword();
```

Comment:
It seems we should throw "unknown user" here instead of "invalid password" to allow `MultipleAccessStorage::loginImpl()` to continue its login attempts with other storages (which follow `LDAPAccessStorage`).

Reply:
...or change back to doing `find()`. :) Having "invalid password" is beneficial for debugging at user side, because this should eventually go to server log.

Reply:
This change would be counter intuitive though, so maybe it's better to add a parameter kind of `bool & continue_login_attempts` to `IAccessStorage::login()` instead.

Reply:
`bool throw_unknown_user_on_invalid_password = false`?

Reply:
The problem of making it a part of common interface is that other storages should also implement this, which is not what we want here.

Reply:
Also, the invalid password are rarely will be communicated as `false`, but rather some exception with `LDAP_ERROR` code. So the solution with `find()` was really best.

Reply:
> Also, invalid passwords are rarely communicated as `false`, but rather some exception with `LDAP_ERROR` code. So the solution with `find()` was really the best.

Yeah, perhaps. But it's counter-intuitive again...

Reply:
Will revert to that.

Reply:
Done.